appinventor/blocklyeditor/src/workspace_svg.js

```diff
@@ -610,12 +646,69 @@ Blockly.WorkspaceSvg.prototype.customContextMenu = function(menuOptions) {
    * @returns {number} returns 0 if the blocks are equal, and -1 or 1 if they are not
    */
   function sortByCategory(a,b) {
-    var comparatorA = comparisonName(a).toLowerCase();
-    var comparatorB = comparisonName(b).toLowerCase();
+    // Sort by Category First, also handles other floating blocks
+    if (a.category == b.category && a.category != "Component"){
+      var comparatorA = comparisonName(a).toLowerCase();
+      var comparatorB = comparisonName(b).toLowerCase();
+      return compareStrTextNum(comparatorA, comparatorB);
+    }
+
+    // 1. Category Global defs first, lexicographically sorted
+    if (a.category == "Variables"){
+      return -1;
+    }
+
+    if (b.category == "Variables"){
+      return +1;
+    }
+
+    // 2. Category Procedure defs next, lexicographically sorted
+    if (a.category == "Procedures"){
+      return -1;
+    }
+
+    if (b.category == "Procedures"){
+      return +1;
+    }
+
+    // 3.Component event handlers, lexicographically sorted by 
+    // type name, instance name, then event name
+    if (a.category == "Component" && b.category == "Component" && a.eventName && b.eventName){
+      if (a.typeName == b.typeName && a.instanceName == b.instanceName &&
```

Comment:
Some future proofing for #1548, you should check if the event has `isGeneric` then it should be sorted before any event with a valid `instanceName` (e.g., not `undefined`)